File: java-simple/src/main/java/dojo/liftpasspricing/Prices.java
package dojo.liftpasspricing;

import static spark.Spark.after;
import static spark.Spark.get;
import static spark.Spark.port;
import static spark.Spark.put;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.List;

public class Prices {

    public static Connection createApp() throws SQLException {

        final Connection connection = DriverManager.getConnection("jdbc:mysql://localhost:3306/lift_pass", "root", "mysql");

        port(4567);

        put("/prices", (req, res) -> {
            int liftPassCost = Integer.parseInt(req.queryParams("cost"));
            String liftPassType = req.queryParams("type");

            try (PreparedStatement stmt = connection.prepareStatement( //
                    "INSERT INTO base_price (type, cost) VALUES (?, ?) " + //
                            "ON DUPLICATE KEY UPDATE cost = ?")) {
                stmt.setString(1, liftPassType);
                stmt.setInt(2, liftPassCost);
                stmt.setInt(3, liftPassCost);
                stmt.execute();
            }

            return "";
        });

        get("/prices", (req, res) -> {
            final Integer age = req.queryParams("age") != null ? Integer.valueOf(req.queryParams("age")) : null;

            int basePrice;
            try (PreparedStatement costStmt = connection.prepareStatement( //
                    "SELECT cost FROM base_price " + //
                            "WHERE type = ?")) {
                costStmt.setString(1, req.queryParams("type"));
                try (ResultSet result = costStmt.executeQuery()) {
                    result.next();
                    basePrice = result.getInt("cost");
                }
            }

            if (age != null && age < 6) {
                return "{ \"cost\": 0}";
            } else {

                if (!req.queryParams("type").equals("night")) {
                    DateFormat isoFormat = new SimpleDateFormat("yyyy-MM-dd");

                    List<Date> holidaysDates = new ArrayList<>();
                    try (PreparedStatement holidayStmt = connection.prepareStatement( //
                            "SELECT * FROM holidays")) {
                        try (ResultSet holidays = holidayStmt.executeQuery()) {

                            while (holidays.next()) {
                                holidaysDates.add(holidays.getDate("holiday"));
                            }
                        }
                    }

                    int reduction = 0;
                    boolean isHoliday = false;
                    for (Date holiday : holidaysDates) {
                        if (req.queryParams("date") != null) {
                            Date d = isoFormat.parse(req.queryParams("date"));
                            if (d.getYear() == holiday.getYear() && //
                                    d.getMonth() == holiday.getMonth() && //
                                    d.getDate() == holiday.getDate()) {
                                isHoliday = true;
                            }
                        }
                    }

                    if (req.queryParams("date") != null) {
                        Calendar calendar = Calendar.getInstance();
                        calendar.setTime(isoFormat.parse(req.queryParams("date")));
                        if (!isHoliday && calendar.get(Calendar.DAY_OF_WEEK) == 2) {
                            reduction = 35;
                        }
                    }

                    // TODO apply reduction for others
                    if (age != null && age < 15) {
                        return "{ \"cost\": " + (int) Math.ceil(basePrice * .7) + "}";
                    } else {
                        if (age == null) {
                            double cost = basePrice * (1 - reduction / 100.0);
                            return "{ \"cost\": " + (int) Math.ceil(cost) + "}";
                        } else {
                            if (age > 64) {
                                double cost = basePrice * .75 * (1 - reduction / 100.0);
                                return "{ \"cost\": " + (int) Math.ceil(cost) + "}";
                            } else {
                                double cost = basePrice * (1 - reduction / 100.0);
                                return "{ \"cost\": " + (int) Math.ceil(cost) + "}";
                            }
                        }
                    }
                } else {
                    if (age != null && age >= 6) {
                        if (age > 64) {
                            return "{ \"cost\": " + (int) Math.ceil(basePrice * .4) + "}";
                        } else {
                            return "{ \"cost\": " + basePrice + "}";
                        }
                    } else {
                        return "{ \"cost\": 0}";
                    }
                }
            }
        });

        after((req, res) -> {
            res.type("application/json");
        });

        return connection;
    }

}

File: cpp/Prices.cpp
#include "Prices.h"
#include <cppconn/resultset.h>
#include <cppconn/statement.h>
#include <iomanip>
#include <string>

#define QUERY_PARAMS(type) (*(req.params.find((type)))).second.c_str()

void Prices::createApp(httplib::Server& server, sql::Connection* con)
{

    server.Put("/prices", [con](const httplib::Request& req, httplib::Response& res) {
        int liftPassCost = std::stoi(QUERY_PARAMS("cost"));
        std::string liftPassType = QUERY_PARAMS("type");

        std::string query = "INSERT INTO lift_pass.base_price (type, cost) "
            "VALUES ('"+ liftPassType + "', " + std::to_string(liftPassCost) + ") "
            "ON DUPLICATE KEY UPDATE cost = " + std::to_string(liftPassCost);

        sql::Statement* stmt = con->createStatement();
        stmt->execute(query);
        con->commit();
        stmt->close();
        delete stmt;

        res.set_content("", "application/json");
    });

    server.Get("/prices", [con](const httplib::Request& req, httplib::Response& res) {
        int age = -1;
        auto it = req.params.find("age");
        if (it != req.params.end()) {
            age = std::stoi((*it).second.c_str());
        }

        std::string liftPassType = QUERY_PARAMS("type");
        std::string query = "SELECT cost FROM lift_pass.base_price "
                            "WHERE type = '" + liftPassType + "'";
        sql::Statement* stmt = con->createStatement();
        sql::ResultSet* rset = stmt->executeQuery(query);
        rset->next();
        int result = rset->getInt("cost");
        delete rset;
        delete stmt;

        int reduction;
        bool isHoliday = false;

        if (age != -1 && age < 6) {
            res.set_content("{ \"cost\": 0}", "application/json");
        }
        else {
            reduction = 0;

            if (strcmp(QUERY_PARAMS("type"), "night") != 0) {
                std::string isoFormat = "%Y-%m-%d";

                std::string query = "SELECT holiday FROM lift_pass.holidays";
                sql::Statement* holidayStmt = con->createStatement();
                sql::ResultSet* holidays = holidayStmt->executeQuery(query);
                while (holidays->next()) {
                    std::string holiday = holidays->getString("holiday");

                    it = req.params.find("date");
                    if (it != req.params.end()) {
                        std::string date = (*it).second.c_str();
                        if (holiday == date) {
                            isHoliday = true;
                        }
                    }
                }
                holidays->close();
                delete holidays;
                holidayStmt->close();
                delete holidayStmt;

                it = req.params.find("date");
                if (it != req.params.end()) {
                    std::tm d = {};
                    std::istringstream ss((*it).second.c_str());
                    ss >> std::get_time(&d, isoFormat.c_str());
                    std::mktime(&d);
                    if (!isHoliday && d.tm_wday == 1) {
                        reduction = 35;
                    }
                }

                /* TODO apply reduction for others */
                if (age != -1 && age < 15) {
                    res.set_content("{ \"cost\": " + std::to_string((int)std::ceil(result * .7)) + "}", "application/json");
                }
                else {
                    if (age == -1) {
                        double cost = result * (1 - reduction / 100.0);
                        res.set_content("{ \"cost\": " + std::to_string((int)std::ceil(cost)) + "}", "application/json");
                    }
                    else {
                        if (age > 64) {
                            double cost = result * .75 * (1 - reduction / 100.0);
                            res.set_content("{ \"cost\": " + std::to_string((int)std::ceil(cost)) + "}", "application/json");
                        }
                        else {
                            double cost = result * (1 - reduction / 100.0);
                            res.set_content("{ \"cost\": " + std::to_string((int)std::ceil(cost)) + "}", "application/json");
                        }
                    }
                }
            }
            else {
                if (age != -1 && age >= 6) {
                    if (age > 64) {
                        res.set_content("{ \"cost\": " + std::to_string((int)std::ceil(result * .4)) + "}", "application/json");
                    }
                    else {
                        res.set_content("{ \"cost\": " + std::to_string(result) + "}", "application/json");
                    }
                }
                else {
                    res.set_content("{ \"cost\": 0}", "application/json");
                }
            }
        }
    });
}
